### DISCUSSION TITLE
Trigger by pull request or merge

### DISCUSSION THREAD

**Discussion by klinux** [terrakube-io/terrakube#405] [Q&A] [answered]
The documentation says: 
**Version Control workflow: Store your Terraform configuration in a git repository, and trigger runs based on pull requests and merges**

But when I create a pull request, the trigger not working, terrakube supports that?

Thank you

**Comment by alfespa17** [comment]
There is an open issue about that feature. https://github.com/AzBuilder/terrakube/issues/396

There is only one small POC using GitHub Actions here https://github.com/AzBuilder/terrakube-action-github

We would like to add that feature in the near future, I think we only need to add the webhooks logic because we already have most of the logic in the current code.

By the way we are open to hear suggestion or ideas of how this kind of featue could be implemented.

**Reply by klinux** [reply]
Thank you for the reply, I believe that feature is very important to provide gitops flows, like Atlantis, with webhooks, it's looks simple, since the  internal logic already done, I think terrakube needs to support webhook for every workspace with VCS support, and listen for pull requests and merges, but I think that you already this mapped in your task, how can I help in this tasks?

**Reply by alfespa17** [reply]
I think the firts step should be to do a POC and add the information inside the issue https://github.com/AzBuilder/terrakube/issues/396

In a high level the idea should be something like this: 

1- Setup some git repository in github with a webhook for example to https://pipedream.com/ just to test.
2- Check the payload in different operations like pull request, merge, etc
3- Validate what information inside the payload depeding of the operation we could use to create the workspaces inside terrakube (check API calls) or any other logic.
4- Validate the type of permission terrakube will need in the repository where the webhook is enabled
5- Define the logic that Terrakube should implement
6- Update the api source code or even create a new component that handle all the webhooks calls

I guess it should be something similar to what atlantis is doing here
https://www.runatlantis.io/docs/configuring-webhooks.html#github-github-enterprise

This could be like the high level idea 👍

**Reply by alfespa17** [reply]
If you would like to help doing the research/POC or even update the code of how this could be implemented all help is welcome 👍

**Comment by jcanizalez** [comment]
hi @klinux @alfespa17 for 2.15.0 I will implement this issue #36 tha basically will trigger the job for each commit in the configured branch in Terrakube. Initially I will implement it for Github.

**Comment by jcanizalez** [comment, marked as answer]
Hi @klinux @alfespa17 
I have implemented webhooks for github  #597  and gitlab #598  will be available as beta next week in version 2.18.0 in case you want to try
